Per-User Credentials for MCP Servers

twinity1 opened the pull request:
## Summary

This PR adds support for user-specific credentials for MCP servers. It allows LibreChat administrators to define custom user variables in the configuration that can be set individually by each user and then referenced in MCP server headers.

Currently, when teams use MCP servers that require authentication, they often use shared/generic credentials for the entire team, which is not ideal from a security and accountability perspective. While LibreChat does support passing the `{{LIBRECHAT_USER_ID}}` variable in headers, this only works for MCP servers specifically designed to handle LibreChat user IDs.

This enhancement allows users to provide their own credentials for third-party MCPs.

## Change Type

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## Testing

- [x] Edit MCP variables from Panel
- [x] Edit MCP variables from MCPSelect
- [x] Test revoke from side panel
- [x] Test revoke from MCPSelect
- [x] Header variables fill to SSE/streamable
- [x] Do not show MCP Settings in the right panel when no servers with custom variables are defined
- [x] Headers/connected are/is refreshed after editing the MCP settings
- [x] Retest web search install/uninstall of variables
- [x] Retest web search using assistant chat
- [x] Add new MCP tool to an agent with dialog (variables are not required for MCP in this case)
- [x] Add new non-MCP tool to an agent with dialog with API Key - tested with OpenWeather and API Key form popup

### **Test Configuration**:
Any MCP server that supports SSE/Streamable HTTP and requires some headers.

## Checklist

- [x] My code adheres to this project's style guidelines
- [x] I have performed a self-review of my own code
- [x] I have commented in any complex areas of my code
- [] My changes do not introduce new warnings
- [x] I have tested the feature with multiple MCP server configurations
- [x] Local unit tests pass with my changes

## Implementation Details

The implementation adds:

1. A new `customUserVars` section in the MCP server configuration:
```yaml
mcpServers:
  redmine:
    url: https://example.com/sse
    headers:
      REDMINE-URL: https://example.com
      REDMINE-API-KEY: "{{REDMINE_API_KEY}}"
    customUserVars:
      REDMINE_API_KEY:
        title: "Redmine API Key"
        description: "Go to https://example.com/my/api_key to get your API key"
```

Variables can be used in headers as shown in the example.

2. UI components for configuring user variables:
    - Users can configure their specific values for these variables in two ways:
        - **From Assistant Chat Input**: A new settings icon appears next to MCP servers in the `MCPSelect.tsx` dropdown (used when selecting MCPs for an assistant). Clicking this icon opens the configuration modal.
        - **From User Settings Panel**: A dedicated "MCP Settings" section in the user settings side panel (`client/src/components/SidePanel/MCP/MCPPanel.tsx`) lists all MCP servers that have `customUserVars` defined by the admin. Clicking a server name here also opens the configuration modal.
    - **Configuration Modal (`MCPConfigDialog.tsx`)**: Both access points trigger the same reusable modal dialog. This dialog dynamically renders input fields based on the `customUserVars` schema defined by the administrator for the selected MCP server. It allows users to input and save their specific values. Features include:
        - Displaying titles and descriptions for each variable as defined in `librechat.yaml`.
        - Secure input fields (e.g., password type) with an option to reveal/hide the value.
    - **Secure Storage**: User-provided variables are sent to the backend via an API call and stored securely (uses existing pluginauths), associated with the user and the specific MCP plugin key (e.g., `mcp_serverName`).

<img width="1020" alt="Screenshot 2025-06-10 at 21 00 24" src="https://github.com/user-attachments/assets/19867858-a138-4fea-b950-ba01e10cd67a" />

<img width="1511" alt="Screenshot 2025-06-10 at 21 00 38" src="https://github.com/user-attachments/assets/acd81def-05b6-40aa-970b-92d7e1ba3f1b" />
<img width="419" alt="Screenshot 2025-06-10 at 21 00 31" src="https://github.com/user-attachments/assets/b4021f6c-9001-46b5-9b3f-4e72c8d71f5a" />

twinity1 commented:
@danny-avila I would like to know your opinion about the idea/implementation before finishing all the tests, code polish, and other requirements for the PR. Thank you!

roman-doubrava commented:
@twinity1 & @danny-avila thank you very much for making this possible. This will enable a lot of things !

twinity1 committed:
chore: add aider to gitignore

twinity1 committed:
feat: fill custom variables to MCP server

twinity1 committed:
feat: replace placeholders with custom user MCP variables

twinity1 committed:
feat: handle MCP install/uninstall (uses pluginauths)

twinity1 committed:
feat: add MCP custom variables dialog to MCPSelect

twinity1 committed:
feat: add MCP custom variables dialog to the side panel

twinity1 committed:
feat: do not require to fill MCP credentials for in tools dialog

twinity1 committed:
feat: add translations keys (en+cs) for custom MCP variables

danny-avila commented:
> Adding credentials from ToolsDialog 

Thanks for looking into it, don't worry about it, I think it's better through the "global" panel anyway

> Group level should not be a problem to implement; I don't see any conflict
> OAuth - I think this one will need a different mechanism and I'm not sure how it should be implemented
Already working on both these items

> The PR's MCP panel is in the agent's section, but I need it to be at the global level, therefore I made at least editing directly in the panel so the UI style is somewhat consistent

Great will take a look may do some minor edits here


---


Download the following extensions on vs code and/or cursor to help resolve ESLint issues:

https://marketplace.visualstudio.com/items?itemName=rvest.vs-code-prettier-eslint
https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint
https://marketplace.visualstudio.com/items?itemName=bradlc.vscode-tailwindcss

not needed but recommended:
https://marketplace.visualstudio.com/items?itemName=eamodio.gitlens
most* of the ESLint errors, you can use "fix all auto-fixable" command from command palette 
once those extensions are installed

---

Please also resolve failing tests and unused i18n keys

twinity1 committed:
fix: handle LIBRECHAT_USER_ID correctly during MCP var replacement

twinity1 committed:
style: remove unused MCP translation keys

twinity1 committed:
style: fix eslint for MCP custom vars

twinity1 commented:
@danny-avila Unfortunately, I'm using WebStorm. But I've fixed it with the ESLint command, so I hope it will be enough.

The i18n and unit tests are also fixed.

danny-avila commented:
Thanks for applying the fixes. Main thing missing now is documentation update to https://github.com/LibreChat-AI/librechat.ai

twinity1 commented on packages/data-provider/src/mcp.ts:
```diff
@@ -204,7 +203,14 @@ export function processMCPEnv(obj: Readonly<MCPOptions>, user?: TUser, customUse
 
-    // 2. Replace standard user field placeholders (now includes USER_ID via ALLOWED_USER_FIELDS)
+    // 2.A. Special handling for LIBRECHAT_USER_ID placeholder
+    // This ensures {{LIBRECHAT_USER_ID}} is replaced only if user.id is available.
```
It was working. I just refactored it into a single function, now that there are custom vars too.

twinity1 commented:
@danny-avila PR with docs: https://github.com/LibreChat-AI/librechat.ai/pull/335

twinity1 commented:
@danny-avila okay, thank you.
